cherab/core/math/integrators/integrators1d.pyx

```diff
@@ -0,0 +1,213 @@
+
+        self.integrand = integrand
+
+        self.relative_tolerance = relative_tolerance
+
+    @property
+    def min_order(self):
+        """
+        Minimum order on Gaussian quadrature.
+
+        :rtype: int
+        """
+        return self._min_order
+
+    @min_order.setter
+    def min_order(self, int value):
+
+        if value < 1:
+            raise ValueError("Order of Gaussian quadrature must be >= 1.")
+
+        if value > self._max_order:
+            raise ValueError("Minimum order of Gaussian quadrature must be less than or equal to the maximum order.")
+
+        self._min_order = value
+
+        self._build_cash()
+
+    @property
+    def max_order(self):
+        """
+        Maximum order on Gaussian quadrature.
+
+        :rtype: float
+        """
+        return self._max_order
+
+    @max_order.setter
+    def max_order(self, int value):
+
+        if value < 1:
+            raise ValueError("Order of Gaussian quadrature must be >= 1.")
+
+        if value < self._min_order:
+            raise ValueError("Maximum order of Gaussian quadrature must be greater than or equal to the minimum order.")
+
+        self._max_order = value
+
+        self._build_cash()
+
+    @property
+    def relative_tolerance(self):
+        """
+        Iteration stops when relative error between last two iterates is less than this value.
+
+        :rtype: double
+        """
+        return self._rtol
+
+    @relative_tolerance.setter
+    def relative_tolerance(self, double value):
+
+        if value <= 0:
+            raise ValueError("Relative tolerance must be positive.")
+
+        self._rtol = value
+
+    cdef _build_cash(self):
+        """
+        Caches the roots and weights of the Gauss-Legendre quadrature.
+        """
+
+        cdef:
+            int order, n, i
+
+        n = (self._max_order + self._min_order) * (self._max_order - self._min_order + 1) // 2
+
+        self._roots = np.zeros(n, dtype=np.float64)
+        self._weights = np.zeros(n, dtype=np.float64)
+
+        i = 0
+        for order in range(self._min_order, self._max_order + 1):
+            self._roots[i:i + order], self._weights[i:i + order] = roots_legendre(order)
+            i += order
+
+        self._roots_mv = self._roots
+        self._weights_mv = self._weights
+
+    @cython.boundscheck(False)
+    @cython.wraparound(False)
+    @cython.cdivision(True)
+    @cython.initializedcheck(False)
+    cdef double evaluate(self, double a, double b) except? -1e999:
+        """
+        Integrates a one-dimensional function over a finite interval.
+
+        :param double a: Lower limit of integration.
+        :param double b: Upper limit of integration.
+
+        :returns: Gaussian quadrature approximation to integral.
+        """
+
+        cdef:
+            int order, i, ibegin
+            double newval, oldval, error, x, c, d
+
+        if self.function is None:
+            raise AttributeError("Integrand is not set.")
```

Comment:
I would keep the function as a property. The integration is going to be called many times during observations

Reply:
The performance of the Cython code will not suffer because `evaluate` is called directly in Cython, and the wrapping is done only in the Python `__call__` method. But never mind, let's leave the integrand as a property.

The problem is that we can't make it mandatory to set the integrand on initialisation. This is so because an instance of `Integrator1D` is passed to the `LineShapeModel` constructor as an argument, but the integrand cannot be set prior `LineShapeModel` initialisation. If setting the integrand to `None` is a bad option, then it should have some default value, for example `Constant1D(0)`.